File: katt/mastermind/output_validators/validator.py
#!/usr/bin/python3

# tester for the task "mastermind"
# usage: ./tester.py input_file correct_output output_dir < contestants_output

import os
from sys import stdin, exit, argv
import sys
import re
import random

def die(msg):
    print(msg)
    f = open(argv[3] + os.sep + "score.txt", "wt+", encoding="utf-8")
    f.write("0")
    f.close()
    exit(43)

def accept(score, real_score):
    f = open(argv[3] + os.sep + "score.txt", "wt+", encoding="utf-8")
    f.write(str(score))
    f.close()
    f = open(argv[3] + os.sep + "judgemessage.txt", "wt+", encoding="utf-8")
    f.write(str(real_score))
    f.close()
    exit(42)

def safe_print(n):
    try:
        print(n)
        sys.stdout.flush()
    except IOError:
        pass

fin, fcor = open(argv[1],'r'), open(argv[2],'r')

best = int(fcor.read())

n, case, case_score, seed = fin.read().split()
n = int(n)
case_score = int(case_score)
seed = int(seed)
assert case in ['exp', '2n', 'n', 'rel']

random.seed(seed)

def score(guesses, best):
    if case == 'exp':
        return 1 if guesses <= 2**n else 0
    if case == '2n':
        return 1 if guesses <= 2*n else 0
    if case == 'n':
        return 1 if guesses <= 2+n else 0
    if case == 'rel':
        assert guesses >= best
        return (n - guesses) / (n - best) if guesses < n else 0
    assert False, "unknown type " + case

target = ''.join(random.choice(['0', '1']) for _ in range(n))

try:
    safe_print(n)

    guesses = 0
    guess = None
    while guess != target:
        guesses += 1
        guess = stdin.readline().strip()
        if len(guess) != n:
            die("wrong length")
        if any(x != '0' and x != '1' for x in guess):
            die("invalid char in guess")
        safe_print(sum(1 for i in range(n) if guess[i] == target[i]))

    sc = score(guesses, best)
    if sc == 0:
        die("too many guesses")
    accept(case_score * sc, guesses)
except ValueError as e:
    die("parse failure: " + e)

File: katt/Contest3/iterated/output_validators/validator.cpp
// usage: ./a.out input_file correct_output output_dir < contestants_output

#include <bits/stdc++.h>
using namespace std;

#define rep(i, from, to) for (int i = from; i < int(to); ++i)
#define sz(x) (int)(x).size()

string output_dir, type;

void die(const string& msg) {
	cout << msg << endl;
	ofstream(output_dir + "/score.txt") << 0;
    exit(43);
}

void accept(double score, int real_score) {
	ofstream(output_dir + "/score.txt") << fixed << setprecision(2) << score;
	ofstream(output_dir + "/judgemessage.txt") << real_score;
    exit(42);
}

void safe_print(int n) {
	try {
		cout << n << endl;
	} catch(...) {}
}

int N;
double score(int guesses, int best) {
	if (type == "exp") {
		assert(N <= 30);
		return (guesses <= (1 << N));
	}
	if (type == "2n") return (guesses <= 2*N);
	if (type == "n") return (guesses <= 2+N);
	if (type == "rel") {
		if (guesses >= N) return 0;
        assert(guesses >= best);
		int a = N - guesses, b = N - best;
		return a / (double)b;
	}
	assert(false);
}

int main(int argc, char** argv) {
	if (argc < 3) exit(1);
	cin.sync_with_stdio(0);
	cin.tie(0);
	int max_score, seed;
	ifstream(argv[1]) >> N >> type >> max_score >> seed;
	output_dir = argv[2];
	srand(seed);

	string guess, target(N, '0');
	rep(i,0,N) target[i] = (char)('0' + ((rand() >> 24) & 1));

	safe_print(N);
	int guesses = 0;
	while (guess != target) {
        guesses++;
		try {
			getline(cin, guess);
			if (!cin) throw 0;
		} catch(...) {
			die("no input!");
		}
		if (sz(guess) != N)
            die("wrong length");
        int s = 0;
		rep(i,0,N) {
			char x = guess[i];
            if (x != 48 && x != 49)
                die("invalid char in guess");
			if (x == target[i])
				s++;
		}
        safe_print(s);
	}
    double sc = score(guesses, N / 2);
    if (sc == 0)
        die("too many guesses");
    accept(max_score * sc, guesses);
}
